Handle RPC requests without response - Closes #16

MaciejBaj opened the pull request:
Add dedicated response cleaners for WAMPClient and SlaveWAMPServer

Closes #16

MaciejBaj committed:
create RequestCleaner instance

MaciejBaj committed:
Implement Client and Internal request cleaners

MaciejBaj committed:
use request cleaners in WAMPClient and SlaveWAMPServer

MaciejBaj committed:
move interval and timeout as constructor parameters

willclarktech commented on SlaveWAMPServer.js:
```diff
@@ -11,16 +12,23 @@ const v = new Validator();
+	constructor(worker,
+		internalRequestsTimeoutMs = 10000,
+		cleanRequestsIntervalMs = 2000,
+		configuredCb = () => {}) {
```
I personally find it much nicer to close brackets on a new line to make it easier to see the difference between e.g. arguments and function body.
```js
constructor(
    worker,
    internalRequestsTimeoutMs = 10000,
    cleanRequestsIntervalMs = 2000,
    configuredCb = () => {},
) {
    super();
    // ...
}
```

willclarktech commented on SlaveWAMPServer.js:
```diff
@@ -11,16 +12,23 @@ const v = new Validator();
-	constructor(worker, configuredCb = () => {}) {
+	constructor(worker,
+		internalRequestsTimeoutMs = 10000,
+		cleanRequestsIntervalMs = 2000,
```
Consider using `e3`, `e6` etc to make it easier to read numbers over `1000`.

willclarktech commented on common/ClientRequestCleaner.js:
```diff
@@ -0,0 +1,24 @@
+
+	constructor(calls, clearInterval, timeout) {
+		super(calls, clearInterval, timeout);
+	}
```
Unnecessary constructor function.

MaciejBaj committed:
apply eslint rules on requests cleaners

willclarktech commented on common/RequestsCleaner.js:
```diff
@@ -0,0 +1,46 @@
+		return timeElapsed > timeout;
+	}
+
+	/* eslint class-methods-use-this: 0 */
```
`/* eslint-disable-next-line class-methods-use-this */`

willclarktech commented on common/RequestsCleaner.js:
```diff
@@ -0,0 +1,46 @@
+
+	/* eslint class-methods-use-this: 0 */
+	verifySignatures() {
+		throw new Error('getSignature needs to be overwritten');
```
`Class method 'verifySignature' must be overwritten by the child.`?

willclarktech commented on common/RequestsCleaner.spec.js:
```diff
@@ -0,0 +1,132 @@
+		beforeEach(() => {
+			clearInterval(requestsCleaner.cleanInterval);
+			requestsCleaner.cleanInterval = null;
+		});
```
Prefer placing `beforeEach` after `before` since `before` hook is run first.

MaciejBaj replied:
right

willclarktech commented on common/RequestsCleaner.spec.js:
```diff
@@ -0,0 +1,132 @@
+
+		describe('when cleanInterval running', () => {
+			beforeEach(() => {
+				if (requestsCleaner.cleanInterval) {
```
This condition is never met.

MaciejBaj replied:
correct

willclarktech commented on common/RequestsCleaner.spec.js:
```diff
@@ -0,0 +1,132 @@
+			it('stop invoking verifySignatures', (done) => {
+				requestsCleaner.stop();
+				setTimeout(() => {
+					expect(requestsCleaner.verifySignatures.called).to.be.false();
```
Don't you have a race condition here if you start the cleaner before the test?

MaciejBaj replied:
because the `verifySignatures` function is being called in the interval it should never be called in this point

willclarktech replied:
If the timeout was set to a much smaller number then this test might fail non-deterministically. A better approach would be to get the `callCount` after being stopped and then assert that it hasn't increased after the timeout.

willclarktech commented on common/RequestsCleaner.spec.js:
```diff
@@ -0,0 +1,132 @@
+	before(() => {
+		requestsCleaner = new RequestsCleaner(validCalls, validIntervalMs, validTimeoutMs);
+	});
+
```
Test for `verifySignatures` required.

willclarktech commented on common/InternalRequestsCleaner.js:
```diff
@@ -0,0 +1,21 @@
+		if (!this.calls) {
+			return;
+		}
+		Object.keys(this.calls).forEach((socketId) => {
```
Consider using the babel polyfill for access to `Object.entries` and `Object.values`.

MaciejBaj replied:
I don't see any advantages here

willclarktech replied:
It's a more semantic way to say what's going on:
```js
Object.values(this.calls).forEach((call) => {
  Object.values(call).forEach((procedure) => {
    Object.entries(procedure).forEach(([signature, callback]) => {
      if (RequestsCleaner.isOutdated(signature, this.timeoutMs)) {
        callback('RPC response timeout exceeded');
        delete call[signature];
      }
    });
  });
});
```

willclarktech replied:
Beyond the scope of this PR as it's a significant decision to add `babel-polyfill` to the project

MaciejBaj replied:
I will stay with not using babel right now

willclarktech commented on common/InternalRequestsCleaner.js:
```diff
@@ -0,0 +1,21 @@
+				Object.keys(this.calls[socketId][procedure]).forEach((signature) => {
+					if (RequestsCleaner.isOutdated(signature, this.timeoutMs)) {
+						this.calls[socketId][procedure][signature]('RPC response timeout exceeded');
+						delete this.calls[socketId][procedure][signature];
```
Some of this deep property access can be avoided by using variables. E.g.
```js
(socketId) => {
    const call = this.calls[socketId];
    Object.keys(call).forEach(...);
}
```
(Also in other files.)

willclarktech commented on common/InternalRequestsCleaner.spec.js:
```diff
@@ -0,0 +1,62 @@
+					},
+				},
+			});
+		});
```
These tests are not atomic.

MaciejBaj replied:
addressed

willclarktech commented on common/RequestsCleaner.js:
```diff
@@ -0,0 +1,46 @@
+	 * @param {number} timeout
+	 * @returns {boolean}
+	 */
+	static isOutdated(signature, timeout) {
```
Do you really want to pass the timeout every time? Will it ever not be `this.timeoutMs`?

willclarktech commented on common/ClientRequestsCleaner.spec.js:
```diff
@@ -0,0 +1,62 @@
+				[validProcedure]: {
+					[validSignature]: validSignatureResolvers,
+				},
+			});
```
These tests are not atomic.

willclarktech commented on SlaveWAMPServer.js:
```diff
@@ -11,16 +12,23 @@ const v = new Validator();
 		this.config = {};
-
+		this.clientRequestsCleaner = new InternalRequestCleaner(
+			this.interProcessRPC, cleanRequestsIntervalMs, internalRequestsTimeoutMs);
```
Likewise, for ease of reading I prefer
```js
this.clientRequestsCleaner = new InternalRequestCleaner(
    this.interProcessRPC, cleanRequestsIntervalMa, internalRequestsTimeoutMs,
);
```
But this is a matter of opinion.

MaciejBaj replied:
it's not a valid code because of comma at the end

willclarktech replied:
This point was about closing the bracket on the next line for clarity.

jondubois commented on common/ClientRequestCleaner.js:
```diff
@@ -0,0 +1,19 @@
+		Object.keys(this.calls).forEach((procedure) => {
+			Object.keys(this.calls[procedure]).forEach((signature) => {
+				if (RequestsCleaner.isOutdated(signature, this.timeoutMs)) {
+					this.calls[procedure][signature].reject('RPC response timeout exceeded');
```
What's your current error and exception handling strategy?

In this specific case a string is OK because you don't really need a stack trace for that, BUT if it is possible for other parts of the code to call the `reject` and pass in `Error` instance (instead of a string) then the handler will now have to deal with two different potential error types (both strings and Error objects) which is inconsistent and adds complexity to the error handlers themselves.

Sometimes strings are not ideal because you want a simple way to identify errors (like an `error.name` or `error.code`) when doing comparisons. Relying on the message can be problematic in the long run: `if (error == 'RPC response timeout exceeded')` is not as nice/futureproof as `if (error.code == 'RPC_RESPONSE_TIMEOUT')`. Node.js itself is moving towards adding a `code` property to internal error objects to decouple the error message from its error code.

It's important that the type of error object/string passed to the `reject` is consistent. If you decide to go with strings all the time, then you need to be sure that they are sufficient for your needs (it will be hard to change in the future once your code starts relying on error messages). There are ways to dehydrate errors to pass them across process boundaries and reconstruct the stack trace (if that is useful at all?). Some projects like to just throw/emit/reject using Objects instead of Errors or strings because they are easier to pass across client/server and process boundaries.

Just something to think about.

MaciejBaj replied:
Thank you very much for the suggestion. I've opened a separate issue to address it #21

MaciejBaj committed:
make requestCleaner.isOutdated method static
- use scientific notation for numbers > 1000
- make request cleaner tests atomic
- include built sources

willclarktech commented on common/ClientRequestsCleaner.spec.js:
```diff
@@ -0,0 +1,64 @@
+			validCalls = {
+				[validProcedure]: {
+					[outdatedSignature]: outdatedSignatureResolvers,
+					[validSignature]: validSignatureResolvers,
```
You need new spies each time.

MaciejBaj replied:
done

willclarktech commented on common/InternalRequestsCleaner.spec.js:
```diff
@@ -0,0 +1,64 @@
+				[validSocketId]: {
+					[validProcedure]: {
+						[outdatedSignature]: outdatedSignatureCallback,
+						[validSignature]: validSignatureCallback,
```
New spies needed

MaciejBaj replied:
done

willclarktech commented on common/RequestsCleaner.spec.js:
```diff
@@ -0,0 +1,135 @@
+	const validIntervalMs = 1;
+
+	before(() => {
+		requestsCleaner = new RequestsCleaner(validCalls, validIntervalMs, valid10secsTimeout10secs);
```
Prefer `beforeEach` for atomicity

MaciejBaj committed:
change default timeouts and replace before with beforeEach

willclarktech commented on common/RequestsCleaner.js:
```diff
@@ -0,0 +1,45 @@
+
+	// eslint-disable-next-line class-methods-use-this
+	verifySignatures() {
+		throw new Error('getSignature needs to be overwritten');
```
This error message needs to be improved

MaciejBaj committed:
improve verifySignatures not implemented error message and stop calling test
